source/grid/grid_tools.cc

```diff
@@ -429,6 +429,140 @@ namespace GridTools
 
 
 
+  // Generic functions for appending face data in 2D or 3D. TODO: we can
+  // remove these once we have 'if constexpr'.
+  namespace
+  {
+    void
+    append_face_data(const CellData<0> & /*face_data*/,
+                     SubCellData & /*subcell_data*/)
+    {
+      Assert(false, ExcInternalError());
+    }
+
+    void
+    append_face_data(const CellData<1> &face_data, SubCellData &subcell_data)
+    {
+      subcell_data.boundary_lines.push_back(face_data);
+    }
+
+    void
+    append_face_data(const CellData<2> &face_data, SubCellData &subcell_data)
+    {
+      subcell_data.boundary_quads.push_back(face_data);
+    }
+  } // namespace
+
+
+
+  template <int dim, int spacedim>
+  std::
+    tuple<std::vector<Point<spacedim>>, std::vector<CellData<dim>>, SubCellData>
+    get_coarse_mesh_description(const Triangulation<dim, spacedim> &tria)
+  {
+    Assert(1 <= tria.n_levels(),
+           ExcMessage("The input triangulation must be non-empty."));
+
+    std::tuple<std::vector<Point<spacedim>>,
+               std::vector<CellData<dim>>,
+               SubCellData>
+      result;
+
+    std::vector<Point<spacedim>> &vertices     = std::get<0>(result);
+    std::vector<CellData<dim>> &  cells        = std::get<1>(result);
+    SubCellData &                 subcell_data = std::get<2>(result);
```

Comment:
too complicated -- just create these three arrays as local variables, and at the end do
```
  return { std::move(vertices), std::move(cells), std::move(subcell_data) };
```
I suspect you might even get away without the `std::move` call, but it won't hurt.

Reply:
Done

Reply:
So I guess the return using just
```
return { std::move(vertices), std::move(cells), std::move(subcell_data) };
```
didn't work without specifying the type?